crates/sdk/benches/array_difference.rs

```diff
@@ -0,0 +1,39 @@
+use criterion::{Criterion, criterion_group, criterion_main};
+use std::{collections::VecDeque, hint::black_box};
+use surrealdb_core::expr::{Array, Number, Value};
+
+// Current implementation as of https://github.com/surrealdb/surrealdb/pull/6047
+// crates/core/src/expr/array.rs
+#[allow(clippy::mutable_key_type)]
+fn array_difference(first: Array, other: Array) -> Array {
+	let mut out = Array::with_capacity(first.len() + other.len());
+	let mut other = VecDeque::from(other.0);
+	for v in first.into_iter() {
+		if let Some(pos) = other.iter().position(|w| v == *w) {
+			other.remove(pos);
+		} else {
+			out.push(v);
+		}
+	}
+	out.append(&mut Vec::from(other));
+	out
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+	let mut first = Array::new();
+	for i in 0..5000 {
+		first.push(Value::Number(Number::Int(i)));
+		first.push(i.to_string().into());
+	}
+	let mut second = Array::new();
+	for i in 0..2500 {
+		second.push(Value::Number(Number::Int(i)));
+		second.push(i.to_string().into());
+	}
+	c.bench_function("array_difference", |b| {
+		b.iter(|| array_difference(black_box(first.clone()), black_box(second.clone())))
+	});
+}
```

Comment:
The results you showed for VecDeque didn't make sense to me remove for VecDeque and Vec are both roughly `O(n)` so the speedup doesn't make sense.

The issue is that this bench case is a best case scenario for the VecDeque and a worse case scenario for the old implementation. Because all the values that need to be removed are at the start the old implementation needed to copy over the rest of the array with each removal while in the new one it is just the equivalent of `pop_front`. 

I don't think this benchmark is indicative of real world data it would probably be better if you used two arrays with random numbers.

Reply:
@DelSkayn Very true and thanks for the close look, I've changed to random data for those two and am seeing a mild performance increase now. (Ran an extra test with just a Vec with ::with_capacity and didn't see any improvement there so VecDeque is still a small improvement, I think)

Just out of curiosity I tried random data with array::complement and saw the same results so haven't made any changes there.

```
old_array_complement    time:   [178.82 ms 179.65 ms 180.64 ms]
array_complement        time:   [2.6079 ms 2.6167 ms 2.6262 ms]
```